cabot/cabotapp/models.py

```diff
@@ -1146,6 +1173,146 @@ def __unicode__(self):
         return "%s: %s to %s%s" % (self.user.username, self.start, self.end, deleted)
 
 
+class ResultFilter(models.Model):
+    """Filter that matches StatusCheckResults."""
+    class Meta:
+        abstract = True
+
+    status_check = models.ForeignKey(StatusCheck, null=False)
+
+    MATCH_CHECK = 'C'
+    MATCH_TYPE_CHOICES = (
+        (MATCH_CHECK, 'Match check.'),
+    )
+    match_if = models.TextField(max_length=1, null=False, blank=False, default=MATCH_CHECK, choices=MATCH_TYPE_CHOICES)
+
+    def matches_result(self, result):
+        # type: (StatusCheckResult) -> bool
+
+        # status_check must match, regardless of match type
+        if result.status_check_id != self.status_check_id:
+            return False
+
+        if self.match_if == self.MATCH_CHECK:
+            return True
+
+        raise NotImplementedError()
+
+
+class Acknowledgement(ResultFilter):
+    """
+    A filter that matches on a StatusCheck (or subset of a StatusCheck, if matching on tags).
+    An Acknowledgement is considered 'open' if closed_at is set to null.
+
+    Acknowledgements "expire" (auto-close) if the expire_at parameter is set. Expiration is performed in two ways:
+      1) get_acks_matching_check() will not return checks where expire_at <= now.
+      2) A periodic celery task calls ack.close() on checks where expire_at <= now.
+
+    Acknowledgements are also automatically closed when their StatusCheck succeeds at least close_after_successes
+    times (consecutively). This is done by Acknowledgement.close_succeeding_acks(), which is called by StatusCheck.run()
```

Comment:
now that we're running StatusChecks after acks are closed, this could mean statuscheck runs -> closes acks -> runs statuscheck again right? This seems fine since the second statuscheck run shouldn't close acks again but just wanted to double check there's no possibility of infinite loop

Reply:
I see what you mean, good idea. I think we're good for now, but this might be tricky if I use signals to trigger the re-run on ack closed...